massifs/localmassifreader.go

```diff
@@ -0,0 +1,384 @@
+func (r *LocalReader) GetHeadMassif(
+	ctx context.Context, tenantIdentityOrLocalPath string,
+	opts ...ReaderOption,
+) (MassifContext, error) {
+
+	dirEntry, ok := r.cache.GetEntry(tenantIdentityOrLocalPath)
+	if ok {
+		return copyCachedMassifOrErr(dirEntry.ReadMassif(r.cache, uint64(dirEntry.HeadMassifIndex)))
+	}
+
+	directory, err := r.cache.ResolveMassifDir(tenantIdentityOrLocalPath)
+	if err != nil {
+		return MassifContext{}, err
+	}
+	err = r.cache.FindMassifFiles(directory)
+	if err != nil {
+		return MassifContext{}, err
+	}
+	dirEntry, ok = r.cache.GetEntry(directory)
+	if !ok {
+		// It's a bug in FindMassifFiles if this case actually happens in practice.
+		return MassifContext{}, fmt.Errorf("failed to prime directory: %s", directory)
+	}
+	return copyCachedMassifOrErr(dirEntry.ReadMassif(r.cache, uint64(dirEntry.HeadMassifIndex)))
+}
+
+func (r *LocalReader) GetFirstMassif(
+	ctx context.Context, tenantIdentityOrLocalPath string,
+	opts ...ReaderOption,
+) (MassifContext, error) {
+	dirEntry, ok := r.cache.GetEntry(tenantIdentityOrLocalPath)
+	if ok {
+		return copyCachedMassifOrErr(dirEntry.ReadMassif(r.cache, uint64(dirEntry.FirstMassifIndex)))
+	}
+
+	directory, err := r.cache.ResolveMassifDir(tenantIdentityOrLocalPath)
+	if err != nil {
+		return MassifContext{}, err
+	}
+	err = r.cache.FindMassifFiles(directory)
+	if err != nil {
+		return MassifContext{}, err
+	}
+	dirEntry, ok = r.cache.GetEntry(directory)
+	if !ok {
+		// It's a bug in FindMassifFiles if this case actually happens in practice.
+		return MassifContext{}, fmt.Errorf("failed to prime directory: %s", directory)
+	}
+	return copyCachedMassifOrErr(dirEntry.ReadMassif(r.cache, uint64(dirEntry.FirstMassifIndex)))
+}
+
+// GetMassifLocalPath returns the local path for the massif identified by the
+// tenant identity and massif index
+func (r *LocalReader) GetMassifLocalPath(tenantIdentity string, massifIndex uint32) string {
+	return filepath.Join(r.GetReplicaDir(), ReplicaRelativeMassifPath(tenantIdentity, massifIndex))
+}
+
+// GetSealLocalPath returns the local path for the seal identified by the tenant identity and massif index
+func (r *LocalReader) GetSealLocalPath(tenantIdentity string, massifIndex uint32) string {
+	return filepath.Join(r.GetReplicaDir(), ReplicaRelativeSealPath(tenantIdentity, massifIndex))
+}
+
+// GetLazyContext is an optimization for remote massif readers
+// and is therefor not implemented for local massif reader
+func (r *LocalReader) GetLazyContext(
+	ctx context.Context, tenantIdentity string, which LogicalBlob,
+	opts ...ReaderOption,
+) (LogBlobContext, uint64, error) {
+
+	return LogBlobContext{}, 0, fmt.Errorf("not implemented for local storage")
+}
+
+// copyCachedSealOrErr deals with errs from ReadSeal and returns a safe copy
+// this exists to simplify the return logic flow in many methods
+func copyCachedSealOrErr(cached *SealedState, err error) (SealedState, error) {
```

Comment:
minor: i don't think we need `OrErr` in the name? its returns an error?

Reply:
ohh i see, you are passing in an error, and if that error is non nil, we return that error. I don't like this pattern,

The caller should just return the error like normal:

if err != nil {
  return nil/SealedState{}, err
}

Reply:
grumble, ok, but i've used that elsewhere a lot to significantly reduce line noise. Your version is tripple the line count and I don't see its really any clearer once you  notice what is going on

Reply:
the caller of this function needs to handle the error in some way anyway?

```
foo, err := copyCachedSealOrErr(bar, err)
if err != nil {
    // handle err
}
```

so if we keep the `OrErr` pattern, we are handling the error twice always, the first handle is just defer handling it to the caller?

I don't know if i am missing something?

Reply:
oh, no you are right, if we are handling twice its pointless. but I thought in all the cases I used it it was a direct return

Reply:
I'm only advocating use of this pattern for direct return sites. Where you can just do: copyCachedSealOrErr(bar, err) and get a bit of DRY for potentially error prone, but common, function exit cases. So to be justified:
1. the call to fooOrErr needs to be part of a return statement
2. there needs to be multiple places where the code in fooOrErr should be used
3. There needs to be something worth factoring out.

Consistent log messages is one case I've used this before. In this case i'm ensuring the caller has a safely independent copy of the item in the cache. And that anything they do with it can't have side effects on future callers obtaining the same cached entry

If you don't think i'm hitting all those buttons, then yes I can back this out. Also, I could just accept a little more duplication, elevate the err check, and still have a function to do the safe copy.

Reply:
the problem is the error returned is the error passed in without altering.

so just return the error in that case?
```
return cached, err 
```

Reply:
if we are worried about nil pointer dereferance on the copy, we should check for nil on the pointer to cached?

Reply:
I've done this all long hand. To much else going on in this PR already, so fair enough, lets not have un-necessariy controversial things